static/js/components/Loading.js

```diff
@@ -1,27 +1,48 @@
 // @flow
 import React from "react"
+import R from "ramda"
 
-import { anyProcessing, anyError, allLoaded } from "../util/rest"
-
-import type { RestState } from "../flow/restTypes"
-
-export default class Loading extends React.Component {
-  props: {
-    restStates: Array<RestState<*>>,
-    renderContents: Function
-  }
+type LoadingProps = {
+  loaded: boolean,
+  errored: boolean
+}
 
-  render() {
-    const { restStates, renderContents } = this.props
+const withLoading = R.curry((
+  shouldLoadData: (Object, Object) => boolean,
+  loadData: Object => void,
+  LoadedComponent: Class<React.Component<*, *, *>>
+) => {
+  return class extends React.Component {
+    props: LoadingProps
 
-    if (anyProcessing(restStates) || !allLoaded(restStates)) {
-      return <div>Loading</div>
+    componentWillMount() {
+      loadData(this.props)
     }
 
-    if (anyError(restStates)) {
-      return <div>Error loading page</div>
+    componentDidUpdate(prevProps: Object) {
+      if (shouldLoadData(prevProps, this.props)) {
+        loadData(this.props)
+      }
     }
 
-    return renderContents()
+    render() {
+      const { loaded, errored, ...props } = this.props
+
+      if (errored) {
+        return <div className="errored">Error loading page</div>
+      }
+
+      if (!loaded) {
+        return <div className="loading">Loading</div>
+      }
+
+      return (
+        <div className="loaded">
+          <LoadedComponent {...props} />
+        </div>
+      )
+    }
   }
-}
+})
+
+export default withLoading
```

Comment:
Personally I like the `updateData` and `shouldUpdateData` methods being defined as pure functions outside of the classes, this would make it easier to test them separately if they need to have any really intense logic in them, if we needed to do that.

I think you're right that that could work basically just as well, but IMO (and this is subjective) the approach @rhysyngsun has put together so far strikes me as more readable and with less indirection.